de.metas.business/src/main/java/de/metas/customs/process/M_InOut_Create_CustomsInvoice.java

```diff
@@ -0,0 +1,224 @@
+	private int p_C_BPartner_Location_ID;
+
+	@Param(parameterName = "AD_User_ID")
+	private int p_AD_User_ID;
+
+	@Param(parameterName = "IsComplete")
+	private boolean p_IsComplete;
+
+	@Override
+	protected String doIt() throws Exception
+	{
+		final ICurrencyBL currencyBL = Services.get(ICurrencyBL.class);
+		final List<InOutAndLineId> linesToExport = retrieveLinesToExport();
+
+		final Map<ProductId, List<InOutAndLineId>> linesToExportMap = linesToExport
+				.stream()
+				.collect(Collectors.groupingBy(inoutAndLineId -> getProductId(inoutAndLineId))); // we asserted earlier that each HU has a locator
+
+		final BPartnerLocationId bpartnerAndLocationId = BPartnerLocationId.ofRepoId(p_C_BPartner_ID, p_C_BPartner_Location_ID);
+		final UserId userId = UserId.ofRepoId(p_AD_User_ID);
+
+		final CurrencyId currencyId = currencyBL.getBaseCurrencyId(Env.getClientId(), Env.getOrgId());
+
+		final LocalDate invoiceDate = Env.getLocalDate();
+
+		final CustomsInvoice customsInvoice = customsInvoiceService.generateCustomsInvoice(bpartnerAndLocationId, userId, currencyId, linesToExportMap, invoiceDate, p_IsComplete);
+
+		final ImmutableSet<InOutId> exportedShippmentIds = linesToExport.stream()
+				.map(InOutAndLineId::getInOutId)
+				.collect(ImmutableSet.toImmutableSet());
+
+		customsInvoiceService.setCustomsInvoiceToShipments(exportedShippmentIds, customsInvoice);
+
+		return MSG_OK;
+
+	}
+
+	private ProductId getProductId(final InOutAndLineId inoutAndLineId)
+	{
+		final IInOutDAO inOutDAO = Services.get(IInOutDAO.class);
+
+		final InOutLineId shipmentLineId = inoutAndLineId.getInOutLineId();
+		final I_M_InOutLine shipmentLineRecord = inOutDAO.getLineById(shipmentLineId, I_M_InOutLine.class);
+
+		return ProductId.ofRepoId(shipmentLineRecord.getM_Product_ID());
+	}
+
+	@Override
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(@NonNull final IProcessPreconditionsContext context)
+	{
+
+		if (context.getSelectionSize() <= 0)
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNoSelection();
+		}
+
+		final boolean foundAtLeastOneUnregisteredShipment = context.getSelectedModels(I_M_InOut.class).stream()
+				.map(shipmentRecord -> InOutId.ofRepoId(shipmentRecord.getM_InOut_ID()))
+				.filter(this::isValidShipment)
+				.map(inoutId -> retrieveValidLinesToExport(inoutId))
+				.findAny()
+				.isPresent();
+
+		return ProcessPreconditionsResolution.acceptIf(foundAtLeastOneUnregisteredShipment);
+	}
+
+	private boolean isValidShipment(final InOutId shipmentId)
+	{
+		final IDocumentBL documentBL = Services.get(IDocumentBL.class);
+
+		final IInOutBL inOutBL = Services.get(IInOutBL.class);
+		final IInOutDAO inOutDAO = Services.get(IInOutDAO.class);
+
+		final I_M_InOut shipment = inOutDAO.getById(shipmentId);
+
+		if (!shipment.isSOTrx())
+		{
+			return false;
+		}
+
+		if (inOutBL.isReversal(shipment))
+		{
+			return false;
+		}
+
+		if (!documentBL.isStatusCompletedOrClosedOrReversed(shipment.getDocStatus()))
```

Comment:
is it really a requirement that we can't create a customs invoice for a completed shippment??

Reply:
Yes. Confirmed with Mark that only completed and closed inouts are to be exported.

Reply:
Right. I didn't notice I used the wrong method :blush: 
I changed it to
```
if (documentBL.isStatusStrOneOf(shipment.getDocStatus(), DocStatus.Completed.getCode(), DocStatus.Closed.getCode()))
```